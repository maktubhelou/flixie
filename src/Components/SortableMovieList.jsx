import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MovieList from './MovieList';
import GenreList from './GenreList';

const backGroundUrl = require("../images/fancy-pants.jpg")
const upArrow = require('../images/upArrow.png');
const downArrow = require('../images/downArrow.png');
const styles = {
    arrowImageStyle: {
        padding: '0',
        width: '36px',
        backgroundColor: 'transparent',
        verticalAlign: 'middle',
        border: 'none',
    },
    buttonStyle: {
        textTransform: 'uppercase',
        height: '24px',
        background: '#222244',
        color: '#9999aa',
        fontSize: '14px',
        textAlign: 'center',
        border: '1px #444455',
        boxShadow: '0 0 3px #9999aa',
        margin: '8px',
        cursor: 'pointer'
    },
    listStyle: {
        backgroundImage: `url(${backGroundUrl})`,
        backgroundAttachment: 'fixed',
        position: 'relative',
        paddingTop: '40px',
        display: 'flex',
        width: '100%',
        flexFlow: 'row wrap',
        justifyContent: 'center' 
    },
    sortRow: {
        width: '100%',
        float: 'left'
    }
}

class SortableMovieList extends Component {
    constructor(props){
        super(props);
        this.state = {
            sortCriteria: 'title',
            sortOrder: 'asc'
        };

        this.sortList = this.sortList.bind(this);
    }

    sortList = (key, order = this.props.sortOrder) => {

        return function (a, b) {
            if(!a.hasOwnProperty(key) || !b.hasOwnProperty(key)) {
                return 0;
            }
            
        const varA = (typeof a[key] === 'string') ? a[key].toUpperCase() : a[key];
        const varB = (typeof b[key] === 'string') ? b[key].toUpperCase() : b[key];
    
        let comparison = 0;
        if (varA > varB) {
          comparison = 1;
        } else if (varA < varB) {
          comparison = -1;
        }
        return (
            (order === 'desc') ? (comparison * -1) : comparison
        );
      };
    }



    render() {
        if(!this.props.loaded) return (<div className="loader" />)

        this.props.movies.sort(this.sortList(this.state.sortCriteria, this.state.sortOrder));

        return (
        <div>
            <div style={styles.sortRow}>  
                <GenreList setGenre={this.props.setGenre} genres={this.props.genres} sortOrder={this.state.sortOrder} />
                <button style={styles.buttonStyle} onClick={() => this.setState({sortCriteria: 'title'})} selected="selected">Title</button>
                <button style={styles.buttonStyle} onClick={() => this.setState({sortCriteria: 'vote_average'})}>Votes</button>
                <button style={styles.buttonStyle} onClick={() => this.setState({sortCriteria: 'genres_id[0]'})}>Genre</button>
                <button style={styles.buttonStyle} onClick={() => this.setState({sortCriteria: 'release_date'})}>Date</button>
                <button onClick={this.toggleSortOrder} style={styles.arrowImageStyle}>{this.state.sortOrder === 'asc' ? <img src={upArrow} style={styles.arrowImageStyle} alt="uparrow"></img> : <img src={downArrow} style={styles.arrowImageStyle} alt="downArrow"></img>}</button>
            </div>
            <div style={styles.listStyle}>
                <MovieList movies={this.props.movies} />
            </div>
        </div>);
    }
}

SortableMovieList.propTypes = {
    movies: PropTypes.array,
    loaded: PropTypes.bool,
    setGenre: PropTypes.func,
    genres: PropTypes.array,
};

export default SortableMovieList;