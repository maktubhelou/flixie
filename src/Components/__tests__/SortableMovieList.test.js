import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import SortableMovieList from '../SortableMovieList'

const fakeList = [
  {
    id: 1,
    title: "One Coder Flew Over the Cuckoo's Nest",
    voteCount: 1245,
    popularity: 567,
    score: 9.9,
    poster_path: 'image.jpg',
    releaseDate: '2019-05-22',
    overview:
      'A story of how one coder lost his mind trying to read his old code.',
  },
]

describe('<SortableMovieList />', () => {
  it('should render without crashing if no data', () => {
    const div = document.createElement('div')
    ReactDOM.render(<SortableMovieList />, div)
    ReactDOM.unmountComponentAtNode(div)
  })
  it('should render without crashing if data', () => {
    const div = document.createElement('div')
    ReactDOM.render(<SortableMovieList movies={fakeList} loaded />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('matches snapshot with data', () => {
    const tree = renderer
      .create(<SortableMovieList movies={fakeList} loaded />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('matches snapshot without data', () => {
    const tree = renderer.create(<SortableMovieList />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
