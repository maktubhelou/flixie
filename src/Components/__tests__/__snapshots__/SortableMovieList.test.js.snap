// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SortableMovieList /> matches snapshot with data 1`] = `
<div>
  <div
    style={
      Object {
        "float": "left",
        "width": "100%",
      }
    }
  >
    <form>
      <select
        style={
          Object {
            "backgroundColor": "rgba(255,255,255,0.7)",
            "border": "10px solid rgba(255,255,255,0.5)",
          }
        }
      />
    </form>
    <button
      onClick={[Function]}
      selected="selected"
      style={
        Object {
          "background": "#222244",
          "border": "1px #444455",
          "boxShadow": "0 0 3px #9999aa",
          "color": "#9999aa",
          "cursor": "pointer",
          "fontSize": "14px",
          "height": "24px",
          "margin": "8px",
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
      type="button"
    >
      Title
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "background": "#222244",
          "border": "1px #444455",
          "boxShadow": "0 0 3px #9999aa",
          "color": "#9999aa",
          "cursor": "pointer",
          "fontSize": "14px",
          "height": "24px",
          "margin": "8px",
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
      type="button"
    >
      Votes
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "background": "#222244",
          "border": "1px #444455",
          "boxShadow": "0 0 3px #9999aa",
          "color": "#9999aa",
          "cursor": "pointer",
          "fontSize": "14px",
          "height": "24px",
          "margin": "8px",
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
      type="button"
    >
      Genre
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "background": "#222244",
          "border": "1px #444455",
          "boxShadow": "0 0 3px #9999aa",
          "color": "#9999aa",
          "cursor": "pointer",
          "fontSize": "14px",
          "height": "24px",
          "margin": "8px",
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
      type="button"
    >
      Date
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "border": "none",
          "padding": "0",
          "verticalAlign": "middle",
          "width": "36px",
        }
      }
      type="button"
    >
      <img
        alt="uparrow"
        src="upArrow.png"
        style={
          Object {
            "backgroundColor": "transparent",
            "border": "none",
            "padding": "0",
            "verticalAlign": "middle",
            "width": "36px",
          }
        }
      />
    </button>
  </div>
  <div
    style={
      Object {
        "backgroundAttachment": "fixed",
        "backgroundImage": "url(fancy-pants.jpg)",
        "display": "flex",
        "flexFlow": "row wrap",
        "justifyContent": "center",
        "paddingTop": "40px",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="border-rounded"
      style={
        Object {
          "backgroundColor": "rgba(49, 49, 66, 0.8)",
          "color": "#eeeeee",
          "height": "100%",
          "margin": "10px",
          "maxWidth": "400px",
          "padding": "10px",
          "textAlign": "center",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "#111133",
            "borderBottom": "1px solid #444455",
            "boxShadow": "0px 4px 16px 0px #111133",
            "color": "#bbbbcc",
            "display": "inline-block",
            "marginBottom": "16px",
            "marginLeft": "-16px",
            "marginTop": "-16px",
            "minHeight": "48px",
            "opacity": "0.9",
            "paddingBottom": "8px",
            "paddingTop": "8px",
            "textAlign": "center",
            "width": "108%",
          }
        }
      >
        <h1
          style={
            Object {
              "paddingLeft": "16px",
            }
          }
        >
          One Coder Flew Over the Cuckoo's Nest
        </h1>
      </div>
      <img
        alt=""
        className="border-rounded"
        name={1}
        style={
          Object {
            "border": "1px solid #444455",
            "boxShadow": "0 0 25px 0 #333355",
          }
        }
      />
      <div
        style={
          Object {
            "color": "#888899",
            "height": "150px",
            "overflowY": "scroll",
            "padding": "10px",
            "textAlign": "justify",
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "#111133",
              "float": "left",
              "marginBottom": "8px",
              "padding": "8px",
            }
          }
        >
          <strong>
            Release Date: 
          </strong>
          2019-05-22
        </div>
        <div
          style={
            Object {
              "backgroundColor": "#111133",
              "float": "right",
              "marginBottom": "8px",
              "padding": "8px",
            }
          }
        >
          <strong>
            Vote average: 
          </strong>
          9.9
        </div>
        <div
          style={
            Object {
              "backgroundColor": "#111133",
              "float": "left",
              "marginBottom": "8px",
              "padding": "8px",
            }
          }
        >
          <strong>
            Vote count: 
          </strong>
          1245
        </div>
        <div
          style={
            Object {
              "backgroundColor": "#111133",
              "float": "right",
              "marginBottom": "8px",
              "padding": "8px",
            }
          }
        >
          <strong>
            Popularity: 
          </strong>
          567
        </div>
        <div
          style={
            Object {
              "clear": "both",
              "marginBottom": "8px",
              "padding": "8px",
            }
          }
        >
          A story of how one coder lost his mind trying to read his old code.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SortableMovieList /> matches snapshot without data 1`] = `
<div
  className="loader"
/>
`;
